server/src/output_stream.rs

```diff
@@ -0,0 +1,121 @@
+    /// usually either ChildStdout or ChildStderr.
+    pub fn setup<R>(stream_type: OutputStream, pipe: R) -> Arc<Self>
+    where
+        R: tokio::io::AsyncRead + Unpin + Send + 'static,
+    {
+        use futures_util::StreamExt;
+
+        let arc_self = Arc::new(OutputHandler::new(stream_type));
+
+        let inner = arc_self.clone();
+        tokio::spawn(async move {
+            let mut out = tokio_util::io::ReaderStream::new(pipe);
+            while let Some(value) = out.next().await {
+                let x = value.expect("Process output error");
+                inner.push(x.to_vec()).await;
+            }
+            inner.complete().await;
+        });
+
+        arc_self
+    }
+
+    /// Push new data to the history, notifying all waiting processes
+    pub async fn push(&self, data: Vec<u8>) {
+        let mut state = self.state.write().await;
+        assert!(
+            !state.completed,
+            "Trying to push more output to a completed stream"
+        );
+        state.history.push(data);
+        self.notify.notify_waiters();
+    }
+
+    /// Mark the process as complete. `push` must not be called after this.
+    pub async fn complete(&self) {
+        let mut state = self.state.write().await;
+        state.completed = true;
+        self.notify.notify_waiters();
+    }
+}
+
+/// Start a task that streams from an `OutputHandler` to a mpsc channel.
+pub fn stream_to(from: Arc<OutputHandler>, to: Sender<Result<OutputEvent, tonic::Status>>) {
+    use std::borrow::Borrow;
+    tokio::spawn(async move {
+        let mut index = 0;
+        loop {
+            let h: &OutputHandler = from.borrow();
+            let (output_if_any, completed): (Option<Vec<u8>>, bool) = {
+                let state = h.state.read().await;
+                let data = if index < state.history.len() {
+                    Some(state.history[index].clone())
+                } else {
+                    None
+                };
+                (data, state.completed)
+            };
+
+            if let Some(output) = output_if_any {
+                let send_result = to
+                    .send(Ok(OutputEvent {
+                        stream: h.stream_type as i32,
+                        output,
+                    }))
+                    .await;
+
+                if send_result.is_err() {
+                    // Send failed, meaning that the other end has hung up.
+                    // In this case it doesn't make sense to stream any more
+                    // output, but this is not an error either.
+                    log::debug!("Stream receiver has hung up, ending stream");
+                    break;
+                }
+
+                index += 1;
+            } else if completed {
+                // All content has been streamed
+                break;
+            } else {
+                // Wait until more output is available
+                h.notify.notified().await;
```

Comment:
What happens if data is pushed to the history between dropping the read lock and registering the waiting task in notify... and then there isn't new data / a new notification for awhile?

Reply:
That causes a delay until more output is pushed. In worst case the client registers after the job has completed, causing the client to wait indefinitely.

This is clearly a bug. Fixed in 2fafbda.

Unfortunately, [`tokio::sync`](https://docs.rs/tokio/1.7.0/tokio/sync/index.html) seems to offer no ready solutions to this. Instread, I rolled my own synchronization method, internally using an atomic generation number, and a `oneshot`-channel based update-notification registration.